### PR TITLE
e-ink ssd1680 driver

### PR DESCRIPTION
**Pull request by adamkaliszan**
## Summary
E-ink driver for chipsed sssd 1680
No BSP files

## Impact
None
## Testing
On ESP32 lilygo tty t5v2.2

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: drivers/lcd/ssd1680.c [hunk 1]
```diff
@@ -0,0 +1,1382 @@
...
+
+static const struct lcd_planeinfo_s g_planeinfo =
+{
+  .putrun  = ssd1680_putrun,             /* Put a run into LCD memory */
+  .putarea = NULL,                       /* Not need to implement */
+  .getrun  = ssd1680_getrun,             /* Read content of shadow memory */
+  .getarea = NULL,                       /* Not need to implement */
+  .buffer  = (FAR uint8_t *)g_runbuffer, /* Run scratch buffer */
+  .bpp     = SSD1680_DEV_BPP,            /* Bits-per-pixel */
+};
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:242]: C89 incompatible

File: drivers/lcd/ssd1680.c [hunk 2]
```diff
@@ -0,0 +1,1382 @@
...
+ *
+ ****************************************************************************/
+
+static int ssd1680_putrun(fb_coord_t row, fb_coord_t col,
+    FAR const uint8_t *buffer, size_t npixels)
+{
+  FAR struct ssd1680_dev_s *priv = (FAR struct ssd1680_dev_s *)&g_epaperdev;
+
+  uint8_t *dst = priv->shadow_fb +
+      row * SSD1680_DEV_ROWSIZE + (col >> SSD1680_PDF);
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:304]: Need to add `FAR`

File: drivers/lcd/ssd1680.c [hunk 3]
```diff
@@ -0,0 +1,1382 @@
...
+ *     0: full off
+ *     CONFIG_LCD_MAXPOWER: full on
+ *   On backlit LCDs, this setting may correspond to the backlight setting.
+ *
+ ****************************************************************************/
+
+static int ssd1680_setpower(FAR struct lcd_dev_s *dev, int power)
+{
+  int ret = OK;
+  struct ssd1680_dev_s *priv = (struct ssd1680_dev_s *) dev;
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:436]: need to add `FAR`

File: drivers/lcd/ssd1680.c [hunk 4]
```diff
@@ -0,0 +1,1382 @@
...
+ * Name:  ssd1680_getcontrast
+ *
+ * Description:
+ *   Get the current contrast setting (0-CONFIG_LCD_MAXCONTRAST).
+ *
+ ****************************************************************************/
+
+static int ssd1680_getcontrast(struct lcd_dev_s *dev)
+{
+  struct ssd1680_dev_s *priv = (struct ssd1680_dev_s *)dev;
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:510]: ditto

File: drivers/lcd/ssd1680.c [hunk 5]
```diff
@@ -0,0 +1,1382 @@
...
+ * Name:  ssd1680_setcontrast
+ *
+ * Description:
+ *   Set LCD panel contrast (0-CONFIG_LCD_MAXCONTRAST).
+ *
+ ****************************************************************************/
+
+static int ssd1680_setcontrast(struct lcd_dev_s *dev, unsigned int contrast)
+{
+  struct ssd1680_dev_s *priv = (struct ssd1680_dev_s *)dev;
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:527]: ditto

File: drivers/lcd/ssd1680.c [hunk 6]
```diff
@@ -0,0 +1,1382 @@
...
+
+/****************************************************************************
+ * Name:  ssd1680_configuredisplay
+ *
+ * Description:
+ *   Setup LCD display.
+ *
+ ****************************************************************************/
+
+static int ssd1680_configuredisplay(struct ssd1680_dev_s *priv)
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:552]: ditto

File: drivers/lcd/ssd1680.c [hunk 7]
```diff
@@ -0,0 +1,1382 @@
...
+ *
+ * Input Parameters:
+ *   priv   - Reference to private driver structure
+ *
+ * Assumptions:
+ *   Caller has selected the OLED section.
+ *
+ ****************************************************************************/
+
+static int ssd1680_redrawfb(struct ssd1680_dev_s *priv)
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:671]: ditto

File: drivers/lcd/ssd1680.c [hunk 8]
```diff
@@ -0,0 +1,1382 @@
...
+  ssd1680_snd_cmd_with_data0(priv, SSD1680_MASTER_ACTIVATE);
+  ssd1680_busy_wait(priv);
+
+  return OK;
+}
+
+static int ssd1680_redrawfb_row(struct ssd1680_dev_s *priv, int row)
+{
+#if defined(CONFIG_LCD_PORTRAIT) || defined(CONFIG_LCD_RPORTRAIT)
+  uint8_t *src = priv->shadow_fb + row * SSD1680_DEV_ROWSIZE;
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:765]: ditto

File: drivers/lcd/ssd1680.c [hunk 9]
```diff
@@ -0,0 +1,1382 @@
...
+ * Description:
+ *   Enable/Disable SSD1680 SPI CS
+ *
+ ****************************************************************************/
+
+static void ssd1680_select(FAR struct ssd1680_dev_s *priv, bool cs)
+{
+  /* If we are selecting the device */
+
+  if (cs == true)
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:852]: `if (cs)`

File: drivers/lcd/ssd1680.c [hunk 10]
```diff
@@ -0,0 +1,1382 @@
...
+      ssd1680_configspi(priv->spi);
+    }
+
+  /* Select/deselect SPI device */
+
+  SPI_SELECT(priv->spi, SPIDEV_DISPLAY(0), cs);
+
+  /* If we are deselecting the device */
+
+  if (cs == false)
```
Comment (pkarashchenko) [inline drivers/lcd/ssd1680.c:866]: `if (!cs)`

**Pull request merged**